pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -965,7 +966,9 @@ func (m *kubeGenericRuntimeManager) computePodActions(ctx context.Context, pod *
 		// If container does not exist, or is not running, check whether we
 		// need to restart it.
 		if containerStatus == nil || containerStatus.State != kubecontainer.ContainerStateRunning {
-			if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus) {
+			// when Evented PLEG is in use, don't restart
+			// for Generic PLEG, restart as is
+			if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus, !pleg.IsEventedPLEGInUse()) {
```

Comment:
When the container is in `Created `state, the new container with the same name will not be created here, are you assuming that the runtime will start it later? When kubelet call cri function `StartContainer`, the container will be in `Created` state, this will make the container stuck in `Created` state forever.

Reply:
Can we write an e2e test for this scenario?

Is there a simple way to reproduce the `startContainer timeout`  in my env?

Reply:
If this is a problem, how about adding a `timeout` check here to see if the pod is in `CREATED` state for a long time? I will try to update the PR later if I can reproduce it.

Reply:
> Can we write an e2e test for this scenario?
> 
> Is there a simple way to reproduce the `startContainer timeout` in my env?

I added some delay codes in containerd to make the `StartContainer` rpc timeout, then the container stuck in `Created `state.

The patch code in [containerd](https://github.com/containerd/containerd/blob/release/1.7/pkg/cri/server/container_start.go#L57) is  like:
```go
if meta.Config != nil {
		for k,v := range meta.Config.Labels {
			if strings.Compare(k, "io.kubernetes.pod.name") == 0 {
				if strings.Compare(v, "centos7-pod") == 0 {
					if blockStat == 0 {
						blockStat = 1
						time.Sleep(5 * time.Minute)
						return nil, errors.New("timeout")
					}
				}
			}
		}
	}
```

and the pod yaml:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: centos7-pod
  labels:
    name: centos7-pod
spec:
  containers:
  - name: centos7-pod
    image: centos:7
    env:
    - name: GET_HOST_FROM
      value: env
    ports:
    - containerPort: 80
    command:
    - sleep
    args:
    - infinity
  nodeSelector:
    kubernetes.io/hostname: ${your node}
```

Reply:
Thanks.

I use a mock cni to just `sleep` and reproduce the issue that you mentioned. 

/hold

I will try to address this issue.

Reply:
A quick fix can be found with the latest commit https://github.com/kubernetes/kubernetes/pull/122778/commits/f09c70d7aa5244dedda084bfcd67955e29abd97f.

@wxx213 could you help to test with the lastest code? I will try it later.

Reply:
I tested locally with your change in containerd.

With the new commit, 
```
centos7-pod   0/1     ContainerCreating   0               24s
centos7-pod   0/1     RunContainerError   0               2m3s
centos7-pod   1/1     Running             1               2m3s
```

Without the new commit(enable EventedPLEG)
```
centos7-pod   0/1     ContainerCreating   0               12s
centos7-pod   0/1     RunContainerError   0               2m13s
...
centos7-pod   0/1     RunContainerError   0               10m 
```

This fixed the problem that mentioned here.